File: python/src/curious/spec_steering.py
from __future__ import annotations

import re

from curious.spec_sections import extract_spec_section, strip_spec_section
from curious.types import Phase
from curious.workflow_policy import sanitize_steering

AGENT_STEERING_HEADING = "## Agent steering"

PHASE_SUBSECTION: dict[Phase, str | None] = {
    "develop": "Developer",
    "review": "Reviewer",
    "sync": "Sync",
    "overseer": None,
}

NOOP_STEERING = re.compile(
    r"^(none|n/a|—|-|\.\.\.|not applicable|no active steering|no steering needed)$",
    re.I,
)
BOILERPLATE_LINE = re.compile(
    r"^(overseer-maintained|injected into|optional\.|_optional|<!--)",
    re.I,
)


def _is_actionable_steering(text: str) -> bool:
    lines = [ln.strip() for ln in text.split("\n") if ln.strip()]
    substantive = []
    for line in lines:
        body = re.sub(r"^[-*]\s*", "", line).strip()
        if not body or NOOP_STEERING.match(body) or BOILERPLATE_LINE.match(body):
            continue
        if body.startswith("_") and body.endswith("_"):
            continue
        if body.startswith("(") and body.endswith(")"):
            continue
        substantive.append(body)
    return len(substantive) > 0


def _extract_subsection(section: str, title: str) -> str | None:
    pattern = rf"^### {re.escape(title)}\s*$([\s\S]*?)(?=^### |\Z)"
    m = re.search(pattern, section, re.M)
    if not m:
        return None
    trimmed = m.group(1).strip()
    if not trimmed or not _is_actionable_steering(trimmed):
        return None
    return trimmed


def agent_steering_for_phase(spec_body: str, phase: Phase) -> str | None:
    if phase == "overseer":
        return None
    section = extract_spec_section(spec_body, AGENT_STEERING_HEADING)
    if not section:
        return None
    subsection_title = PHASE_SUBSECTION.get(phase)
    if subsection_title:
        phase_specific = _extract_subsection(section, subsection_title)
        if phase_specific:
            text, _ = sanitize_steering(phase_specific)
            return text if text.strip() and _is_actionable_steering(text) else None
        if re.search(r"^### ", section, re.M):
            return None
    if _is_actionable_steering(section):
        text, _ = sanitize_steering(section)
        return text if _is_actionable_steering(text) else None
    return None


def strip_agent_steering(spec_body: str) -> str:
    return strip_spec_section(spec_body, AGENT_STEERING_HEADING)


def format_steering_prompt_block(phase: Phase, steering: str) -> str:
    role = PHASE_SUBSECTION.get(phase) or phase
    return (
        f"## Agent steering (overseer → {role})\n\n"
        "Corrective guidance from the overseer — follow in addition to AGENTS.md and the spec.\n\n"
        f"{steering.strip()}"
    )

File: src/spec-steering.ts
import { extractSpecSection, stripSpecSection } from "./spec-sections.js";
import { sanitizeSteering } from "./workflow-policy.js";
import type { Phase } from "./types.js";

export const AGENT_STEERING_HEADING = "## Agent steering";

const PHASE_SUBSECTION: Partial<Record<Phase, string>> = {
  develop: "Developer",
  review: "Reviewer",
  sync: "Sync",
};

/** Placeholder / empty steering — do not inject into prompts. */
const NOOP_STEERING =
  /^(none|n\/a|—|-|\.\.\.|not applicable|no active steering|no steering needed)/i;

const BOILERPLATE_LINE =
  /^(overseer-maintained|injected into|optional\.|_optional|<!--)/i;

function isActionableSteering(text: string): boolean {
  const lines = text
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line.length > 0);

  const substantive = lines.filter((line) => {
    const body = line.replace(/^[-*]\s*/, "").trim();
    if (!body) return false;
    if (NOOP_STEERING.test(body)) return false;
    if (BOILERPLATE_LINE.test(body)) return false;
    if (body.startsWith("_") && body.endsWith("_")) return false;
    // Template placeholders from examples
    if (body.startsWith("(") && body.endsWith(")")) return false;
    return true;
  });

  return substantive.length > 0;
}

function extractSubsection(section: string, title: string): string | null {
  const pattern = new RegExp(
    `^### ${title}\\s*$([\\s\\S]*?)(?=^### |\\z)`,
    "m",
  );
  const match = section.match(pattern);
  if (!match) return null;
  const trimmed = match[1].trim();
  if (trimmed.length === 0 || !isActionableSteering(trimmed)) {
    return null;
  }
  return trimmed;
}

/** Steering text injected into develop / review / sync prompts. */
export function agentSteeringForPhase(
  specBody: string,
  phase: Phase,
): string | null {
  if (phase === "overseer") {
    return null;
  }

  const section = extractSpecSection(specBody, AGENT_STEERING_HEADING);
  if (!section) {
    return null;
  }

  const subsectionTitle = PHASE_SUBSECTION[phase];
  if (!subsectionTitle) {
    return section;
  }

  const phaseSpecific = extractSubsection(section, subsectionTitle);
  if (phaseSpecific) {
    const { text } = sanitizeSteering(phaseSpecific);
    if (!text.trim() || !isActionableSteering(text)) {
      return null;
    }
    return text;
  }

  // Shared bullets only when there are no ### subsections at all
  const hasSubsections = /^### /m.test(section);
  if (!hasSubsections && isActionableSteering(section)) {
    const { text } = sanitizeSteering(section);
    return isActionableSteering(text) ? text : null;
  }

  return null;
}

export function stripAgentSteering(specBody: string): string {
  return stripSpecSection(specBody, AGENT_STEERING_HEADING);
}

export function formatSteeringPromptBlock(
  phase: Phase,
  steering: string,
): string {
  const role = PHASE_SUBSECTION[phase] ?? phase;
  return [
    `## Agent steering (overseer → ${role})`,
    "",
    "Corrective guidance from the overseer — **only present when something concrete needed improvement**. Follow below in addition to AGENTS.md and the spec; if this block is absent, proceed normally.",
    "",
    steering.trim(),
  ].join("\n");
}
